include/distributed_graph_generator.h

```diff
@@ -0,0 +1,152 @@
+#pragma once
+
+#include <bitset>
+#include <unordered_map>
+#include <variant>
+
+#include "command_graph.h"
+#include "ranges.h"
+#include "region_map.h"
+#include "types.h"
+
+namespace celerity::detail {
+
+class task;
+class task_manager;
+class abstract_command;
+
+// TODO: Make compile-time configurable
+constexpr size_t max_num_nodes = 256;
+using node_bitset = std::bitset<max_num_nodes>;
+
+/**
+ * write_command_state is a command_id with two bits of additional information:
+ *   - Whether the data written by this command is globally still the newest version ("fresh" or "stale")
+ *   - Whether this data has been replicated from somewhere else, i.e., we are not the original producer
+ */
+class write_command_state {
+	constexpr static int64_t fresh_bit = 1ll << 63;
+	constexpr static int64_t replicated_bit = 1ll << 62;
+	static_assert(sizeof(fresh_bit) == sizeof(command_id));
+
+  public:
+	constexpr write_command_state() = default;
+
+	/* explicit(false) */ constexpr write_command_state(command_id cid) : m_cid(cid) {}
+
+	constexpr write_command_state(command_id cid, bool is_replicated) : m_cid(cid) {
+		if(is_replicated) { m_cid |= replicated_bit; }
+	}
+
+	bool is_fresh() const { return (m_cid & fresh_bit) == 0u; }
+
+	bool is_replicated() const { return (m_cid & replicated_bit) != 0u; }
+
+	void mark_as_stale() { m_cid |= fresh_bit; }
+
+	operator command_id() const { return m_cid & ~fresh_bit & ~replicated_bit; }
+
+	bool operator==(const write_command_state& other) const { return m_cid == other.m_cid; }
+
+	bool operator!=(const write_command_state& other) const { return !(*this == other); }
+
+  private:
+	command_id m_cid = 0;
+};
+
+class distributed_graph_generator {
+	friend struct distributed_graph_generator_testspy;
+
+	inline static const write_command_state no_command = write_command_state(static_cast<command_id>(-1));
+
+	struct buffer_state {
+		buffer_state(region_map<write_command_state> lw, region_map<std::bitset<max_num_nodes>> rr)
+		    : local_last_writer(std::move(lw)), replicated_regions(std::move(rr)), pending_reduction(std::nullopt) {}
+
+		region_map<write_command_state> local_last_writer;
+		region_map<node_bitset> replicated_regions;
+
+		// When a buffer is used as the output of a reduction, we do not insert reduction_commands right away,
+		// but mark it as having a pending reduction. The final reduction will then be generated when the buffer
+		// is used in a subsequent read requirement. This avoids generating unnecessary reduction commands.
+		std::optional<reduction_info> pending_reduction;
```

Comment:
Well as you noted we actually do need to access the `local_last_writer` still, so that would then have had to either be replicated into the `pending_reduction_state` or moved to a separate map altogether. Other than that (and implementation order, as you note), I don't think so. In any case, this code isn't going to last anyway.

Reply:
Sure, let's address this in the IDAG reduction overhaul.